update to use openrs2 and fix some bugs I found

ARTARNA opened the pull request.

ARTARNA committed:
update to use openrs2 and fix some bugs I found

ScoreUnder commented on src/main/kotlin/cache/loaders/TextureLoader.kt:
```diff
@@ -36,7 +36,14 @@ class TextureLoader(cacheLibrary: CacheLibrary) {
+            // Return empty array if no textures were loaded successfully
+            return arrayOfNulls(0)
+        }
+        val maxId = textureDefinitionCache.keys.maxOrNull() ?: -1
```
This will never be null as there is a check earlier to see if the texture cache is empty

which btw, should it ever be empty? either we succeed and load some textures, or we shouldn't have called this at all, right?

ScoreUnder replied:
Reverted this because we can load textures from new caches now (#206)

ScoreUnder commented on src/main/kotlin/cache/loaders/TextureLoader.kt:
```diff
@@ -36,7 +36,14 @@ class TextureLoader(cacheLibrary: CacheLibrary) {
-        val maxId = textureDefinitionCache.maxBy { it.key }!!.key
+        if (textureDefinitionCache.isEmpty()) {
+            // Return empty array if no textures were loaded successfully
+            return arrayOfNulls(0)
```
`emptyArray()`?

ScoreUnder replied:
Reverted this because we can load textures from new caches now (#206)

ScoreUnder commented on src/main/kotlin/cache/loaders/TextureLoader.kt:
```diff
@@ -49,35 +56,95 @@ class TextureLoader(cacheLibrary: CacheLibrary) {
-                def.field1780[var3] = inputStream.readUnsignedByte()
-            }
+    private fun load(id: Int, b: ByteArray?): TextureDefinition? {
+        if (b == null || b.isEmpty()) {
```
Why do we want to allow null/empty byte arrays to be passed to this function?

ScoreUnder replied:
Reverted this because we can load textures from new caches now (#206)

ScoreUnder commented on src/main/kotlin/cache/ParamsManager.kt:
```diff
@@ -12,6 +12,11 @@ class ParamsManager {
 
     fun loadFromPath(path: String) {
         val paramsFile = File(path, "params.txt")
+        if (!paramsFile.exists()) {
```
huge nitpick but this is a TOCTOU race, just catch the IO error instead

ScoreUnder replied:
done

ScoreUnder commented on build.gradle.kts:
```diff
@@ -64,6 +64,11 @@ ktlint {
 }
 
+java {
+    sourceCompatibility = JavaVersion.VERSION_19
```
Do we have any Java here?

ScoreUnder commented on src/main/kotlin/cache/loaders/TextureLoader.kt:
```diff
@@ -49,35 +56,95 @@ class TextureLoader(cacheLibrary: CacheLibrary) {
+            // Any other error loading the texture, skip it
+            System.err.println("Warning: Failed to load texture $id: ${e.message}, skipping.")
+            return null
         }
```
a lot of manual checks for remaining buffer + catching in load function. failure is an exceptional condition, it should be fine to let it bubble up and get handled by the caller

ScoreUnder replied:
Reverted this because we can load textures from new caches now (#206)

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -410,16 +551,19 @@ class CacheChooserController(
+            val dateStr = localDate?.format(DateTimeFormatter.ISO_LOCAL_DATE)
 
-            if(localDate == date) {
+            if(dateStr == date?.format(DateTimeFormatter.ISO_LOCAL_DATE)) {
```
why compare dates as strings?

ScoreUnder replied:
Removed the stringification

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -281,53 +306,169 @@ class CacheChooserController(
                         } else {
-                            Files.copy(tarIn, dest.toPath())
+                            null
+                        }
```
val revisionNumber = cache.builds.firstOrNull()?.major

ScoreUnder replied:
Changed

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -281,53 +306,169 @@ class CacheChooserController(
                         }
-                        tarEntry = tarIn.nextTarEntry
+                    } catch (e: Exception) {
+                        println("Warning: Could not create params.txt: ${e.message}")
```
losing type and stack trace?

ScoreUnder replied:
Using logger instead

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -281,53 +306,169 @@ class CacheChooserController(
+                            ).use { writer ->
+                                writer.write("Created at ${java.time.LocalDateTime.now()}\n")
+                                writer.write("\n")
+                                writer.write("codebase=http://oldschool${revisionNumber}.runescape.com/\n")
```
why actively fake a potentially incorrect codebase if we don't use the codebase?

ScoreUnder replied:
Cut down to essentials + comment saying it's not a "real" params file

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -281,53 +306,169 @@ class CacheChooserController(
+                                entryName.substring(6) // Remove "cache/" prefix
+                            } else {
+                                entryName
+                            }
```
val relativePath = entryName.removePrefix("cache/")

ScoreUnder replied:
Done

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -281,53 +306,169 @@ class CacheChooserController(
-                            dest.mkdirs()
+                    // Delete existing cache folder if it exists to avoid conflicts
+                    val cacheDir = File(destFolder, "cache")
+                    if (cacheDir.exists()) {
```
is this check necessary? deleteRecursively shouldn't throw if it doesn't exist

ScoreUnder replied:
Removed check

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -461,16 +605,52 @@ class CacheChooserController(
+            }
+        } else {
+            "?"
+        }
```
done elsewhere too -- might warrant an extension method for `Build` rather than copy-paste?

ScoreUnder replied:
Done

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -281,53 +306,169 @@ class CacheChooserController(
+                                entryName
+                            }
+                            val dest = File(destFolder, "cache/$relativePath")
+                            dest.parentFile?.mkdirs()
```
Have `File(destFolder, "cache")` in another variable, you'll avoid the need for null-checking parentFile and also avoid the need for manual path joining here

ScoreUnder replied:
Since `relativePath` may contain a slash (subdirectories), perhaps this is fine as-is

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -281,53 +306,169 @@ class CacheChooserController(
+                        zipIn.closeEntry()
+                        zipEntry = zipIn.nextEntry
+                    }
+                    zipIn.close()
```
I know the original code didn't do this, but it could benefit from `.use {}` rather than manual closing

ScoreUnder replied:
Done

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -281,53 +306,169 @@ class CacheChooserController(
+            try {
+                val instant = Instant.parse(it)
+                val localDate = instant.atZone(ZoneId.of("UTC")).toLocalDate()
+                localDate.format(DateTimeFormatter.ISO_LOCAL_DATE)
```
why the parse-and-format round trip?

ScoreUnder replied:
Left as-is but abbreviated into an extension property (dateString). Still not sure why it's using this approach though?

ScoreUnder replied:
I guess it's more friendly to present to a user than the full RFC 3339 timestamp, which is the format we get from OpenRs2

ScoreUnder commented on src/main/kotlin/controllers/CacheChooserController.kt:
```diff
@@ -281,53 +306,169 @@ class CacheChooserController(
+        }.start()
     }
 
     private fun downloadCache(
```
this is getting fairly hefty; might be nicer if it had its own class at this point. also it's been a while since I touched RS2 stuff but is there a reason to remove runestats support entirely? are they cancelled? lol

ScoreUnder commented:
Mostly some questions in review. Nothing super problematic jumps out though there are a few things I find confusing. The biggest question is why the RuneStats support is being dropped entirely -- kinda big for a PR with no description!

Could you expand on the bugs you're fixing? I can see some amount of error handling being added in some places, but it's not clear what this does in the grand scheme of things.

ScoreUnder committed:
Revert texture loader changes

Texture loading has been fixed for newer caches -- no longer need to
work around it by skipping the textures

ScoreUnder committed:
Merge branch 'main' into HEAD

ScoreUnder committed:
Simplify some version and string-handling logic

ScoreUnder committed:
Use logger instead of println

ScoreUnder committed:
Whitespace nitpicks

ScoreUnder committed:
Trim down generated params.txt

ScoreUnder committed:
Simplify date logic by factoring out common methods

ScoreUnder committed:
Use File(parent, child) rather than slash directory separator

ScoreUnder committed:
Simplify exception message test

ScoreUnder committed:
Move OpenRs2 API calls into OpenRs2Api

+ Factor out some repeated code
+ Use consistent user agent
+ Avoid URL.openConnection

ScoreUnder commented:
I've resolved some of the easier comments I have on this myself (with apologies: it's on the main branch on your fork, so I've had to push up to there!), now what remains is:

- Why remove RuneStats cache support entirely?
- ~Why parse-and-format timestamp rather than just not parsing it in the first place?~ friendlier than RFC 3339 for presentation
- Might want a refactor to move cache saving logic out of UI controller class
- ~Still want to remove the TOCTOU race just on principle lol~ done

Bigger maybes:

- Might want to save metadata as-is
  - which could also remove the need for a synthetic `params.txt`
- Might want to locally cache the list of available caches to avoid spamming the server if someone restarts the program a lot. Expiry of 12h or something? This is a bit of a longer one tho so understandable to not implement it

I've put in what I have time to put in on this, I'll maybe come back to it when I'm more free and then put some time into some refactoring, or reinstating RuneStats support alongside OpenRs2. Otherwise please discuss in the meantime if you're able!

ScoreUnder committed:
Don't test existence before reading params.txt

- Exceptions are still possible due to a TOCTOU race, so just handle the
  exception either way.
- We get marginally less overhead on the happy path this way (though
  this is unlikely to mean anything in real life)
